src/mapper.js

```diff
@@ -1,39 +1,104 @@
+        break;
+      }
+    }
+  }
+  //return the value for the first mapped item found
+  return value;
+};
+
+const getTag = (mapItems) => {
+  const result = [];
+  for (const Key in mapItems) {
+    const Value = mapItems[Key];
+    result.push({ Key, Value });
+  }
+  if (result.length) {
+    return result;
+  }
+};
+
+const isEmpty = (obj) => {
+  for (const prop in obj) {
+    if (obj.hasOwnProperty(prop)) return false;
+  }
+  return true;
+};
+
+const getSet = (mapItems, properties) => {
+  if (Array.isArray(mapItems)) {
+    let resolvedValues = [];
+    mapItems.forEach((input) => {
+      resolvedValues.push(mapper(input, properties));
+    });
+    return resolvedValues;
+  }
+};
+
+const mapper = (input, schema) => {
   const mappedObject = {};
-  Object.keys(schema).forEach((schemaKey) => {
-    const dataType = schema[schemaKey].type;
-    const mapItems = schema[schemaKey].mapItems;
-    const defaultValue = schema[schemaKey].defaultValue ?? null;
-    switch (dataType) {
-      case "string":
-      case "number":
-      case "boolean":
-      case "array":
-        if(schema[schemaKey].staticValue) {
-          mappedObject[schemaKey] = schema[schemaKey].staticValue;
-          return;
-        }
-        const resolvedPath = mapItems ? mapItems.find(mapItem => get(input, mapItem) !== undefined) : false;
-        let resolvedValue = resolvedPath ? get(input, resolvedPath) : defaultValue;
-        if(schema[schemaKey].mappingValueRules) {
-          const validMapRule = schema[schemaKey].mappingValueRules.find(rule => rule.original.includes(resolvedValue));
-          if(validMapRule) {
-            resolvedValue = validMapRule.target;
-          }
-        }
-        mappedObject[schemaKey] = resolvedValue;
-        return;
-      case "object":
-        mappedObject[schemaKey] = mapper(
-          input,
-          schema[schemaKey].properties
-        );
-        return;
-      // case "array(objects)":
+  for (const key in schema) {
+    const value = schema[key];
+    const dataType = value.type;
+    let mapItems;
+    let resolvedValue;
+    let hasMapItems = false;
+
+    if (value.staticValue) {
+      mappedObject[key] = value.staticValue;
+      continue;
     }
-  });
-  return mappedObject;
-}
+
+    if (value.mapItems) {
+      mapItems = getMapItems(input, value.mapItems);
+      hasMapItems = typeof mapItems !== "undefined";
+    }
+
+    if (hasMapItems) {
+      if (dataType === "set") {
+        resolvedValue = getSet(mapItems, value.properties);
+      } else if (dataType === "tag") {
+        resolvedValue = getTag(mapItems);
+      } else {
+        resolvedValue = value.properties
+          ? mapper(mapItems, value.properties)
+          : mapValueRules(value.mappingValueRules, mapItems);
+      }
+    } else {
+      resolvedValue = value.properties
+        ? mapper(input, value.properties)
+        : value.defaultValue;
+    }
+
+    if (typeof resolvedValue !== "undefined") {
```

Comment:
As discussed with @kim-cloudconformity, items that don't map to a value staticValue, defaultValue will no longer be returned in the transformed object.

Reply:
I'm not sure I understand this comment but we can chat about it tomorrow.

Reply:
We are looping through the schema object, which means properties are included even if there is no value for them, e.g, even with no terraform values for "VersioningConfiguration" and "WebsiteConfiguration", previously we were getting:

```
        "VersioningConfiguration": {
            "Status": null,
            "MFADelete": null
        },
        "WebsiteConfiguration": {
            "IndexDocument": null,
            "ErrorDocument": null,
            "RoutingRules": null
        } 
```
Now, these fields are excluded from the transformed object, unless there is a defaultValue or staticValue

Reply:
Thanks @togleid, I remember now.